src/v/archival/ntp_archiver_service.cc

```diff
@@ -2613,6 +2664,72 @@ ss::future<> ntp_archiver::apply_spillover() {
     }
 }
 
+flush_result ntp_archiver::flush() {
+    // Return early if we are not the leader.
+    if (!_parent.is_leader()) {
+        vlog(_rtclog.debug, "in flush(), not leader.");
+        return flush_result::not_leader;
+    }
+
+    _flush_uploads_offset = model::prev_offset(
+      max_uploadable_offset_exclusive());
+    vlog(
+      _rtclog.debug,
+      "accepted flush(), flush offset is {}.",
+      _flush_uploads_offset.value());
+    return flush_result::accepted;
+}
+
+ss::future<wait_result> ntp_archiver::wait() {
+    // If there hasn't been a flush request, return early.
+    if (!flush_in_progress()) {
+        vlog(_rtclog.debug, "In wait(), flush() is not in progress.");
+        co_return wait_result::not_in_progress;
+    }
+
+    // If our manifest's insync offset is greater than the user provided offset,
+    // we have uploaded or flush'ed() up to this point.
+    if (current_flush_complete()) {
+        vlog(_rtclog.debug, "In wait(), flush() is already complete.");
+        co_return wait_result::complete;
+    }
+
+    // If this partition has lost leadership, we indicate so, and the user
+    // may want to reissue a flush() command to the partition leader.
+    if (lost_leadership()) {
+        vlog(_rtclog.debug, "In wait(), leadership was already lost.");
+        co_return wait_result::lost_leadership;
+    }
+
+    // Otherwise, we will wait on the condition variable to signal.
+    try {
+        vlog(_rtclog.debug, "Waiting for flush() to complete");
+        co_await _flush_cond.wait();
+    } catch (const ss::broken_condition_variable&) {
+        // We shutdown while waiting for a flush() to finish.
+        // Return a failure result.
+        co_return wait_result::failed;
+    }
+
+    // Recheck flush status.
+    if (current_flush_complete()) {
+        vlog(_rtclog.debug, "After waiting in wait(), flush was completed.");
+        _flush_uploads_offset.reset();
+        co_return wait_result::complete;
+    }
+
+    // Recheck leadership. No need to reset _flush_uploads_offset here, it will
+    // be done in the upload_until_abort loop.
+    if (lost_leadership()) {
+        vlog(_rtclog.debug, "After waiting in wait(), leadership was lost.");
+        co_return wait_result::lost_leadership;
+    }
+
+    // Not sure how we got here, but we failed.
+    vlog(_rtclog.debug, "After waiting in wait(), failed to flush.");
+    co_return wait_result::failed;
```

Comment:
I think it's possible to get here if we lose leadership and then become leader again before control is given to this waiter fiber.

We probably want to check after waiting for the condition that there's still a flush in progress, and it not, return `lost_leadership`?

Reply:
`lost_leadership()` considers if the current term differs from the original term of the `_parent` partition- that seems like it covers this case, please correct me if I'm wrong.

Reply:
`lost_leadership()` only tells us whether we've started a new iteration in `upload_until_abort()`, but it doesn't itself track the term `wait()` started in. Consider the following sequence:

1. `wait()` called in term 1, setting the flush offset and we begin waiting on the flush cond
2. `upload_until_abort()` runs, but loses leadership briefly, ending one iteration
3. `upload_until_abort()` regains leadership in term 2, starting a new iteration, resetting the flush offset and signaling the cv
4. note that at this point, `lost_leadership()` will return false because it has started a new iteration in the new term
5. the `wait()` fiber responds to the signal and gets to this `failed` branch

I think it would be more apt if `lost_leadership()` were called `not_leader()`. There's an implicit assumption with "lost leadership" that the caller cares about a specific point in time at which it was leader, but that isn't true in its implementation, which only checks the current leadership status

Reply:
Thanks for the thoughtful response.

I have left `lost_leadership()` as is, but I have changed around some of the logic in `wait()` by storing the `_start_term` in a variable before calling `_flush_cond.wait()`, and comparing that to the `_start_term` post `_flush_cond.wait()` to see if the term has changed.

I think this is better logic, but I won't know for sure until I begin writing some tests with leadership transfer during a `flush()` (coming soon).